decoder/semantic_tokens.go

```diff
@@ -0,0 +1,123 @@
+package decoder
+
+import (
+	"sort"
+
+	"github.com/hashicorp/hcl-lang/lang"
+	"github.com/hashicorp/hcl-lang/schema"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+)
+
+// SemanticTokensInFile returns a sequence of semantic tokens
+// within the config file.
+func (d *Decoder) SemanticTokensInFile(filename string) ([]lang.SemanticToken, error) {
+	f, err := d.fileByName(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := d.bodyForFileAndPos(filename, f, hcl.InitialPos)
+	if err != nil {
+		return nil, err
+	}
+
+	if d.rootSchema == nil {
+		return []lang.SemanticToken{}, nil
+	}
+
+	tokens := tokensForBody(body, d.rootSchema, false)
+
+	sort.Slice(tokens, func(i, j int) bool {
+		return tokens[i].Range.Start.Byte < tokens[j].Range.Start.Byte
+	})
+
+	return tokens, nil
+}
+
+func tokensForBody(body *hclsyntax.Body, bodySchema *schema.BodySchema, isDependent bool) []lang.SemanticToken {
+	tokens := make([]lang.SemanticToken, 0)
+
+	if bodySchema == nil {
+		return tokens
+	}
+
+	for name, attr := range body.Attributes {
+		attrSchema, ok := bodySchema.Attributes[name]
+		if !ok {
+			if bodySchema.AnyAttribute == nil {
+				// unknown attribute
+				continue
+			}
+			attrSchema = bodySchema.AnyAttribute
+		}
+
+		modifiers := make([]lang.SemanticTokenModifier, 0)
+		if isDependent {
+			modifiers = append(modifiers, lang.TokenModifierDependent)
+		}
+		if attrSchema.IsDeprecated {
+			modifiers = append(modifiers, lang.TokenModifierDeprecated)
+		}
+
+		tokens = append(tokens, lang.SemanticToken{
+			Type:      lang.TokenAttrName,
+			Modifiers: modifiers,
+			Range:     attr.NameRange,
+		})
+	}
+
+	for _, block := range body.Blocks {
+		blockSchema, ok := bodySchema.Blocks[block.Type]
+		if !ok {
+			// unknown block
+			continue
+		}
+
+		modifiers := make([]lang.SemanticTokenModifier, 0)
+		if isDependent {
+			modifiers = append(modifiers, lang.TokenModifierDependent)
+		}
+		if blockSchema.IsDeprecated {
+			modifiers = append(modifiers, lang.TokenModifierDeprecated)
+		}
+
+		tokens = append(tokens, lang.SemanticToken{
+			Type:      lang.TokenBlockType,
+			Modifiers: modifiers,
+			Range:     block.TypeRange,
+		})
+
+		for i, labelRange := range block.LabelRanges {
+			if i+1 > len(blockSchema.Labels) {
```

Comment:
Feels odd to me, but I'm not 100% up to speed with what is going on

Reply:
Good catch, there should be `>=`. This is just to catch a situation of e.g. 3rd label declared for `resource` which has schema for just 2 labels.

Reply:
Actually this works just fine as is. This is what happens in that situation:

```go
if i+1 > len(blockSchema.Labels)
```
```go
// 3rd label has index of 2
// schema only declares 2 labels
if 2+1 > 2
```
```go
if 3 > 2 # true
```